sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java

```diff
@@ -341,63 +324,113 @@ public UnsafeArrayData copy() {
-        "it's too big.");
-    }
+  @Override
+  public boolean[] toBooleanArray() {
+    int size = numElements();
+    boolean[] values = new boolean[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.BYTE_ARRAY_OFFSET, size);
+    return values;
+  }
+
+  @Override
+  public byte[] toByteArray() {
+    int size = numElements();
+    byte[] values = new byte[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.BYTE_ARRAY_OFFSET, size);
+    return values;
+  }
+
+  @Override
+  public short[] toShortArray() {
+    int size = numElements();
+    short[] values = new short[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.SHORT_ARRAY_OFFSET, size * 2);
+    return values;
+  }
 
-    final int offsetRegionSize = 4 * arr.length;
-    final int valueRegionSize = 4 * arr.length;
-    final int totalSize = 4 + offsetRegionSize + valueRegionSize;
-    final byte[] data = new byte[totalSize];
+  @Override
+  public int[] toIntArray() {
+    int size = numElements();
+    int[] values = new int[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.INT_ARRAY_OFFSET, size * 4);
+    return values;
+  }
 
-    Platform.putInt(data, Platform.BYTE_ARRAY_OFFSET, arr.length);
+  @Override
+  public long[] toLongArray() {
+    int size = numElements();
+    long[] values = new long[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.LONG_ARRAY_OFFSET, size * 8);
+    return values;
+  }
 
-    int offsetPosition = Platform.BYTE_ARRAY_OFFSET + 4;
-    int valueOffset = 4 + offsetRegionSize;
-    for (int i = 0; i < arr.length; i++) {
-      Platform.putInt(data, offsetPosition, valueOffset);
-      offsetPosition += 4;
-      valueOffset += 4;
+  @Override
+  public float[] toFloatArray() {
+    int size = numElements();
+    float[] values = new float[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.FLOAT_ARRAY_OFFSET, size * 4);
+    return values;
+  }
+
+  @Override
+  public double[] toDoubleArray() {
+    int size = numElements();
+    double[] values = new double[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.DOUBLE_ARRAY_OFFSET, size * 8);
+    return values;
+  }
+
+  private static UnsafeArrayData fromPrimitiveArray(Object arr, int length, final int elementSize) {
+    final int headerSize = calculateHeaderPortionInBytes(length);
+    if (length > (Integer.MAX_VALUE - headerSize) / elementSize) {
+      throw new UnsupportedOperationException("Cannot convert this array to unsafe format as " +
+        "it's too big.");
     }
 
+    final int valueRegionSize = elementSize * length;
+    final byte[] data = new byte[valueRegionSize + headerSize];
```

Comment:
I decided not to change  from  4 bytes to 8 bytes for ```numElements```. This is because ```numElements()``` is defined as ```int``` in ```ArrayData```.
It would be good to create another PR to change a type for ```numElements()```.